tests/cluster/tests/28-cluster-shards.tcl

```diff
@@ -0,0 +1,244 @@
+        }
+        if { $slots == $slot2 } {
+            assert_equal [llength $nodes] 2
+            dict_setup_for_verification primary 2
+            dict_setup_for_verification replica 6
+            dict_shard_cmp [lindex $nodes 0] $primary
+            dict_shard_cmp [lindex $nodes 1] $replica
+            incr validation_cnt
+        }
+        if { $slots == $slot3 } {
+            assert_equal [llength $nodes] 2
+            dict_setup_for_verification primary 1
+            dict_setup_for_verification replica 5
+            dict_shard_cmp [lindex $nodes 0] $primary
+            dict_shard_cmp [lindex $nodes 1] $replica
+            incr validation_cnt
+        }
+        if { $slots == $slot4 } {
+            assert_equal [llength $nodes] 2
+            dict_setup_for_verification primary 0
+            dict_setup_for_verification replica 4
+            dict_shard_cmp [lindex $nodes 0] $primary
+            dict_shard_cmp [lindex $nodes 1] $replica
+            incr validation_cnt
+       }
+    }
+    assert_equal $validation_cnt 4
+}
+
+# Remove the only slot owned by primary 3, slots array should be empty.
+test "Verify no slots shard" {
+    R 3 cluster DELSLOTSRANGE 1001 1001
+    set shards [R 3 CLUSTER SHARDS]
+    set validation_cnt 0
+    foreach shard $shards {
+        set slots [dict get $shard slots]
+        set nodes [dict get $shard nodes]
+        if {[llength $slots] == 0 && [llength $nodes] == 2} {
+                dict_setup_for_verification primary 3
+                dict_setup_for_verification replica 7
+                dict_shard_cmp [lindex $nodes 0] $primary
+                dict_shard_cmp [lindex $nodes 1] $replica
+                incr validation_cnt 1
+        }
+    }
+    assert_equal $validation_cnt 1
+    R 3 cluster ADDSLOTSRANGE 1001 1001
+}
+
+set id0 [R 0 CLUSTER MYID]
+
+set current_epoch [CI 2 cluster_current_epoch]
+
+test "Killing one primary node" {
+    kill_instance redis 0
+}
+
+test "Wait for failover" {
+    wait_for_condition 1000 50 {
+        [CI 2 cluster_current_epoch] > $current_epoch
+    } else {
+        fail "No failover detected"
+    }
+}
+
+test "Cluster should eventually be up again" {
+    assert_cluster_state ok
+}
+
+# Primary 0 node should report as fail.
+test "Verify health as fail for killed node" {
+    set shards [R 1 CLUSTER SHARDS]
+    set validation_cnt 0
+    foreach shard $shards {
+        set slots [dict get $shard slots]
+        set nodes [dict get $shard nodes]
+        set slot_len [llength $slots]
+        if {$slot_len == 0 && [dict get [lindex $nodes 0] id] == $id0} {
+            dict set primary id $id0
+            dict set primary hostname "host-0.com"
+            dict set primary port 30000
+            dict set primary health FAIL
+            dict_shard_cmp [lindex $nodes 0] $primary
+            incr validation_cnt 1
+        }
+    }
+    assert_equal $validation_cnt 1
+}
+
+test "Cluster is writable" {
+    cluster_write_test 1
+}
```

Comment:
Why are there two writable tests?

Reply:
After failover, I've added this to verify if the cluster is writable. Is it unnecessary ?

Reply:
Why wouldn't the cluster be writable? We normally don't test this after each failover